File: src/main/java/me/tomasan7/jecnadesktop/ui/SubPageManager.java
package me.tomasan7.jecnadesktop.ui;

import javafx.scene.Parent;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.Pane;
import me.tomasan7.jecnadesktop.JecnaDesktop;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Is responsible for swapping {@link Page SubPages} in a sub-page container ({@link Pane}).
 * Eagerly loads all {@link CachedPage ChachedPages}.
 */
public class SubPageManager
{
	/** The MainPage SubPage pane. */
	private final Pane subPageContainer;

	private final JecnaDesktop jecnaDesktop;
	/**
	 * Cached {@link Parent Parents} for {@link Page subpages}.
	 */
	private final Map<SubPage, Page> pages = new HashMap<>();

	/** The currently active (viewed) {@link Page subpage} */
	private SubPage currentPage;

	private final Set<SubPageSwitchListener> subPageSwitchListeners = new HashSet<>();

	public SubPageManager (JecnaDesktop jecnaDesktop, Pane subPageContainer)
	{
		this.subPageContainer = subPageContainer;
		this.jecnaDesktop = jecnaDesktop;
	}

	public void switchToPage (SubPage subPage)
	{
		subPageContainer.getChildren().clear();
		Parent subPageParent = pages.get(subPage).getContent();
		subPageContainer.getChildren().add(subPageParent);

		AnchorPane.setLeftAnchor(subPageParent, 0d);
		AnchorPane.setTopAnchor(subPageParent, 0d);
		AnchorPane.setRightAnchor(subPageParent, 0d);
		AnchorPane.setBottomAnchor(subPageParent, 0d);

		SubPage oldSubPage = currentPage;
		currentPage = subPage;

		/* Invoking the listeners. */
		subPageSwitchListeners.forEach(listener -> listener.onSwitch(oldSubPage, subPage));
	}

	/**
	 * Adds a {@link Page} to be handled by the {@link SubPageManager}.
	 * @param subPage The {@link SubPage} enum value.
	 * @param page The {@link Page} to be viewed.
	 */
	public void addSubPage (SubPage subPage, Page page)
	{
		/* This is to eagerly load and cache all cached pages. */
		if (page instanceof CachedPage cachedPage)
			cachedPage.getContent();

		pages.put(subPage, page);
	}

	/**
	 * Removes a {@link Page} to be handled by the {@link SubPageManager}.
	 * @param subPage The {@link SubPage} enum value.
	 */
	public void removeSubPage (SubPage subPage)
	{
		pages.remove(subPage);
	}

	public void registerSubPageSwitchListener (SubPageSwitchListener listener)
	{
		subPageSwitchListeners.add(listener);
	}

	public void unregisterSubPageSwitchListener (SubPageSwitchListener listener)
	{
		subPageSwitchListeners.remove(listener);
	}

	/**
	 * @return The currently active (viewed) {@link Page subpage}.
	 */
	public SubPage getCurrentPage ()
	{
		return currentPage;
	}
}

File: src/main/kotlin/me/tomasan7/jecnadesktop/ui/SubPageManager.kt
package me.tomasan7.jecnadesktop.ui

import javafx.scene.Parent
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.Pane
import java.util.LinkedList

/**
 * Is responsible for swapping sub-pages in a sub-page container ([Pane]).
 * Eagerly loads all [CachedPage]s.
 *
 * @property subPageContainer The MainPage SubPage pane.
 */
class SubPageManager(
    private val subPageContainer: Pane
)
{
    /**
     * Cached [Parent]s for subpages.
     */
    private val pages = HashMap<SubPage, Page>()

    /** The currently active (viewed) subpage. */
    var currentPage: SubPage? = null

    private val subPageSwitchListeners = LinkedList<(SubPage?, SubPage) -> Unit>()

    fun switchToPage (subPage: SubPage)
    {
        subPageContainer.children.clear()
        val subPageParent = pages[subPage]!!.content
        subPageContainer.children.add(subPageParent)

        AnchorPane.setLeftAnchor(subPageParent, 0.0)
        AnchorPane.setTopAnchor(subPageParent, 0.0)
        AnchorPane.setRightAnchor(subPageParent, 0.0)
        AnchorPane.setBottomAnchor(subPageParent, 0.0)

        val oldSubPage = currentPage
        currentPage = subPage

        /* Invoking the listeners. */
        subPageSwitchListeners.forEach { it(oldSubPage, currentPage!!) }
    }

    /**
     * Adds a [Page] to be handled by the [SubPageManager].
     * @param subPage The [SubPage] enum value.
     * @param page The [Page] to be viewed.
     */
    fun addSubPage (subPage: SubPage, page: Page)
    {
        /* This is to eagerly load and cache all cached pages. */
        if (page is CachedPage)
            page.content

        pages[subPage] = page
    }

    /**
     * Removes a [Page] to be handled by the [SubPageManager].
     * @param subPage The [SubPage] enum value.
     */
    fun removeSubPage (subPage: SubPage)
    {
        pages.remove(subPage)
    }

    fun registerSubPageSwitchListener (listener: (SubPage?, SubPage) -> Unit)
    {
        subPageSwitchListeners.add(listener)
    }

    fun unregisterSubPageSwitchListener (listener: (SubPage?, SubPage) -> Unit)
    {
        subPageSwitchListeners.remove(listener)
    }
}
